### ISSUE TITLE
Workaround for tree-sitter test file separator conflict

### ISSUE THREAD

**Issue by ahelwer** [tree-sitter/tree-sitter#982]
Copied from [this](https://stackoverflow.com/q/66606818/2852699) stackoverflow question, since upon examining the source code I don't see a workaround.

TLA+ uses `----` and `====` in the top-level definition of a module:

```
---- MODULE Test ----
====
```

This conflicts with the tree-sitter test file format, so I can't have more than one test in a file. One solution to this issue is to expose a CL parameter for `tree-sitter test` which changes the character used in the separators. For example:

```
tree-sitter test -start "@" -end "%"
```

**Comment by ahlinc** [comment]
I'd suggest to introduce some delimiters changing operator in a test file itself. Something similar to mysql `DELIMITER ...` https://www.mysqltutorial.org/mysql-stored-procedure/mysql-delimiter/ cause command line arguments is ephemeral things and it  would be better that delimiters definitions been commited to VCS with texts of tests.

**Comment by maxbrunsfeld** [comment]
Alternatively, we could have the test parser look at the first "header line" in the test file, and check if it has any optional *suffix* after the `===` signs, and if so, require that *every* separator needs to have that suffix. For example:

```
==============|||
Things
==============|||

---- My ---
------ Source
--- Code--

-----------------|||

(my
  (expected
     (parse)
     (tree)))

=================|||
Other things
=================|||

more
-----
source
----------
code

----------|||

(more
  (parse)
  (trees))
```

**Comment by ahlinc** [comment]
> the test parser look at the first "header line" in the test file, and check if it has any optional suffix after the === signs, and if so, require that every separator needs to have that suffix.

+1 for this idea, it looks similar to MIME boundaries separator.

**Comment by Sjord** [comment]
> TLA+ uses ---- and ==== in the top-level definition of a module:
> 
> ```
> ---- MODULE Test ----
> ====
> ```

This does not necessarily look ambigious to me. I would expect the separator between code and tree to be a line containing just `---` and optionally more dashes, but nothing else. So `---- MODULE Test ----` would not be a separator. The documentation can be interpreted either way:

> Then the input source code is written, followed by a line containing three or more - (dash) characters.

`====` occurs here in the middle of the code, after the title but before the syntax tree. In that position, it should not end the current test but should be considered part of the test code.

So I would argue against changing the format of the tests, and I would prefer if the parsing was a little bit smarter.

**Comment by ahelwer** [comment]
@Sjord kinda seems like kicking the can down the road. No matter what parsing scheme you come up with for the text file there will theoretically exist *some* language construct that makes parsing the file ambiguous.

**Comment by ahelwer** [comment]
For workaround I have created a NPM package built with the PR changes here: https://www.npmjs.com/package/ahelwer-tree-sitter-cli

It can be used until the PR changes are merged and included in a tree-sitter-cli release.

**Comment by ahelwer** [comment]
Released a new version of the package rebased on the 0.20.0 tag: https://www.npmjs.com/package/ahelwer-tree-sitter-cli

**Comment by ahelwer** [comment]
Going forward the npm packages will be served from https://www.npmjs.com/package/@tlaplus/tree-sitter-cli and the https://www.npmjs.com/package/ahelwer-tree-sitter-cli packages will be archived/hidden.

Just released a 0.20.1-0 prerelease package that includes fixes for the `tree-sitter highlight` command.
